File: main/tree-huffman-decoding/tree-huffman-decoding.java
/*  
	class Node
		public  int frequency; // the frequency of this tree
    	public  char data;
    	public  Node left, right;
    
*/ 

	void decode(final String s, final Node root) {
        final StringBuilder sb = new StringBuilder();
        
        Node node = root;
        for (int i = 0; i != s.length(); ++i) {
            node = (s.charAt(i) == '0' ? node.left : node.right);
            
            if (node.left == null && node.right == null) {
                sb.append(node.data);
                node = root;
            }
        }
        
        System.out.println(sb);
    }

File: main/tree-huffman-decoding/tree-huffman-decoding.cpp
/* 
The structure of the node is

typedef struct node
{
    int freq;
    char data;
    node * left;
    node * right;
    
}node;

*/


void decode_huff(node * const root, const string s)
{
    auto cur = root;
    
    for (const auto b : s) {
        cur = (b == '0' ? cur->left : cur->right);
        
        if (cur->data != '\0') {
            cout << cur->data;
            cur = root;
        }
    }
}
